README.md
## Redmine OAuth plugin 2.1.2

[![GitHub CI](https://github.com/onixsib/redmine_oauth/actions/workflows/rubyonrails.yml/badge.svg?branch=main)](https://github.com/onixsib/redmine_oauth/actions/workflows/rubyonrails.yml)

This plugin is used to authenticate in Redmine through an OAuth provider.

The user is identified by the email registered by the OAuth provider. The email must correspond with an email registered
in Redmine. If such an email is not found, the user is ofered with registration to Redmine depending on the Redmine's
setting **Self-registration**.

Inspired by Gucin's plugin https://github.com/Gucin/redmine_omniauth_azure.
Inspired by kontron plugin <https://github.com/kontron/redmine_oauth/>.

Supported OAuth providers:
* Azure AD (https://azure.microsoft.com)
* Otka (https://www.okta.com)
* GitLab (https://about.gitlab.com)
* Keycloak (<https://www.keycloak.org>)

### Installation:

1. Enter the plugins folder
2. Clone the repository
3. Set user permissions
4. Install required gems
5. Restart the application

E.g. Linux + Apache web server

```shell
cd plugins
git clone https://github.com/onixsib/redmine_oauth.git
chown -R www-data:www-data redmine_oauth
cd ..
bundle install
systemctl restart apache2
```

**There is a conflict with *net-imap* gem!**

```
[!] There was an error parsing `Gemfile`: You cannot specify the same gem twice with different version requirements.
You specified: net-imap (~> 0.2.2) and net-imap (~> 0.3.1). Bundler cannot continue.
```

This gem supports XOAUTH from v0.3.1. So, it is necessary to comment out
corresponding line in Redmine's Gemfile and then run `bundle install` again.

```
-- gem 'net-imap', '~> 0.2.2'
++ #gem 'net-imap', '~> 0.2.2'
```

### Registration

Register your Redmine instance as an application by your OAuth provider. Follow the instructions given on their web
sites. As the redirect URI add https://yourdomain/oauth2callback.

### Configuration

Open _Administration -> Plugins_ in your Redmine and configure the plugin.

 E.g.:

**Provider**  Azure AD

**Site**  https://login.microsoftonline.com

**Client ID** xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx

**Client secret** xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

**Tenant ID** xxxxxxxx-xxx-xxxx-xxxx-xxxxxxxxxxxx

### Tasks

## Receive IMAP
Read emails from an IMAP server and process them into Redmine.

Available options:
* host - IMAP server [outlook.office365.com]
* port - Port [993]
* scope - Scope ['https://outlook.office365.com/.default']
* grant_type - Grant type ['client_credentials']
* ssl - use SSL [Yes]
* starttls - Start TLS [No]
* username - Login
* folder - Mail folder to scan [INBOX]
* move_on_success - Where to move successfully processed messages
* move_on_failure - Where to move unsuccessfully processed messages

Example:

```rake redmine_oauth:email:receive_imap username='notifications@example.com' RAILS_ENV="production"```

**Prior accessing IMAP via OAuth, it is necessary to grant flow to authenticate IMAP connections.**

Here is a procedure how to do that in Azure:

https://learn.microsoft.com/en-us/exchange/client-developer/legacy-protocols/how-to-authenticate-an-imap-pop-smtp-application-by-using-oauth#use-client-credentials-grant-flow-to-authenticate-imap-and-pop-connections

### Uninstallation

```
cd plugins
rm redmine_oauth
```
Then restart the application/web server.

CHANGELOG.md
Changelog for Redmine OAuth
==========================

2.1.2 *2023-10-04*
------------------

    Keycloak OAuth provider

2.1.1 *2023-06-02*
------------------

    GiLab enhancement

2.1.0 *2023-06-02*
------------------

    GitLab OAuth provider

* New: #12 - Add support for GitLab OAuth provider

2.0.2 *2023-02-10*
------------------

    Back URL

* Bug: #8 - back_url bug

2.0.1 *2022-12-23*
------------------

    Invalid credentials error

2.0.0 *2022-11-08*
------------------

    Receive IMAP over OAuth rake task

1.0.4 *2022-10-26*
------------------

    Users' last connection timestamp

1.0.3 *2022-10-18*
------------------

    Right processing of locked users

1.0.2 *2022-10-10*
------------------

    German localisation

1.0.1 *2022-10-07*
------------------

    Okta OAuth provider
    Cross Site Request Forgery implementation
    OAuth login button facelift

1.0.0 *2022-09-16*
------------------

    Azure AD OAuth provider


init.rb
# frozen_string_literal: true

# Redmine plugin OAuth
#
# Karel Pičman <karel.picman@kontron.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

require 'redmine'
require File.expand_path('lib/redmine_oauth/hooks', __dir__)

Redmine::Plugin.register :redmine_oauth do
  name 'Redmine OAuth plugin'
  author 'nbm a@onixsib.ru'
  description 'Redmine OAuth plugin with Keycloak'
  version '2.1.2'
  url 'https://github.com/onixsib/redmine_oauth'
  author_url 'https://github.com/onixsib/redmine_oauth/graphs/contributors'
  settings default: {
    oauth_name: 'none',
    site: '',
    client_id: '',
    client_secret: '',
    tenant_id: '',
    button_color: '#ffbe6f',
    button_icon: 'fas fa-address-card'
  }, partial: 'settings/oauth_settings'
end


assets/javascripts/redmine_oauth.js
/*
 Redmine plugin OAuth

 Karel Pičman <karel.picman@kontron.com>

 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 2
 of the License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

function oauth_set_color()
{
    let color = $("input#button_color").val();
    $("#login-oauth-button").css({ backgroundColor: color });
}

function oauth_set_icon()
{
    let icon_class = $("select#settings_button_icon option:selected").val();
    let login_button = $("#login-oauth-button");
    if(icon_class == 'none'){
        login_button.hide();
        return;
    }
    else{
        login_button.show();
    }
    let icon = $("i#button_icon");
    icon.removeClass();
    icon.addClass(icon_class);
}

function oauth_settings_visibility()
{
    let div_oauth_options = $("div#oauth_options");
    let tenant_id = $("input#settings_tenant_id");
    let button = $("button#login-oauth-button");
    let oauth_name = $("#settings_oauth_name option:selected").val();
    let html = button.html();
    $("input#settings_site").val("");
    $("input#settings_client_id").val("");
    $("input#settings_client_secret").val("");
    html = html.replace(/<b>.*<\/b>/, "<b>" + oauth_name + "</b>");
    button.html(html);
    switch(oauth_name) {
        case 'none':
            div_oauth_options.hide();
            tenant_id.val("");
            break;
        case 'Azure AD':
            div_oauth_options.show();
            div_oauth_options.find('#oauth_options_tenant').show();
            tenant_id.val("");
            break;
        case 'Okta':
            div_oauth_options.show();
            div_oauth_options.find('#oauth_options_tenant').show();
            tenant_id.val("default");
            break;
        case 'Keycloak':
            div_oauth_options.show();
            div_oauth_options.find('#oauth_options_tenant').show();
            tenant_id.val("default");
            break;
        case 'GitLab':
            div_oauth_options.show();
            div_oauth_options.find('#oauth_options_tenant').hide();
            break;
        default:
            break;
    }
}

app/controllers/redmine_oauth_controller.rb
# frozen_string_literal: true

# Redmine plugin OAuth
#
# Karel Pičman <karel.picman@kontron.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

require 'account_controller'
require 'jwt'

# OAuth controller
class RedmineOauthController < AccountController
  before_action :verify_csrf_token, only: [:oauth_callback]

  def oauth
    session[:back_url] = params[:back_url]
    oauth_csrf_token = generate_csrf_token
    session[:oauth_csrf_token] = oauth_csrf_token
    case Setting.plugin_redmine_oauth[:oauth_name]
    when 'Azure AD'
      redirect_to oauth_client.auth_code.authorize_url(
        redirect_uri: oauth_callback_url,
        state: oauth_csrf_token,
        scope: 'user:email'
      )
    when 'GitLab'
      redirect_to oauth_client.auth_code.authorize_url(
        redirect_uri: oauth_callback_url,
        state: oauth_csrf_token,
        scope: 'read_user'
      )
    when 'Okta'
      redirect_to oauth_client.auth_code.authorize_url(
        redirect_uri: oauth_callback_url,
        state: oauth_csrf_token,
        scope: 'openid profile email'
      )
    when 'Keycloak'
      redirect_to oauth_client.auth_code.authorize_url(
        redirect_uri: oauth_callback_url,
        state: oauth_csrf_token,
        scope: 'openid profile email offline_access'
      )
    else
      flash['error'] = l(:oauth_invalid_provider)
      redirect_to signin_path
    end
  rescue StandardError => e
    Rails.logger.error e.message
    flash['error'] = e.message
    redirect_to signin_path
  end

  def oauth_callback
    raise StandardError, l(:notice_access_denied) if params['error']

    case Setting.plugin_redmine_oauth[:oauth_name]
    when 'Azure AD'
      token = oauth_client.auth_code.get_token(params['code'], redirect_uri: oauth_callback_url)
      user_info = JWT.decode(token.token, nil, false).first
      email = user_info['unique_name']
    when 'GitLab'
      token = oauth_client.auth_code.get_token(params['code'], redirect_uri: oauth_callback_url)
      userinfo_response = token.get('/api/v4/user', headers: { 'Accept' => 'application/json' })
      user_info = JSON.parse(userinfo_response.body)
      user_info['login'] = user_info['username']
      email = user_info['email']
    when 'Okta'
      token = oauth_client.auth_code.get_token(params['code'], redirect_uri: oauth_callback_url)
      userinfo_response = token.get(
        "/oauth2/#{Setting.plugin_redmine_oauth[:tenant_id]}/v1/userinfo",
        headers: { 'Accept' => 'application/json' }
      )
      user_info = JSON.parse(userinfo_response.body)
      user_info['login'] = user_info['preferred_username']
      email = user_info['email']
    when 'Keycloak'
      token = oauth_client.auth_code.get_token(params['code'], redirect_uri: oauth_callback_url)
      userinfo_response = token.get(
        "/realms/#{Setting.plugin_redmine_oauth[:tenant_id]}/protocol/openid-connect/userinfo",
        headers: { 'Accept' => 'application/json' }
      )
      user_info = JSON.parse(userinfo_response.body)
      user_info['login'] = user_info['preferred_username']
      email = user_info['email']
    else
      raise StandardError, l(:oauth_invalid_provider)
    end
    raise StandardError, l(:oauth_no_verified_email) unless email

    try_to_login email, user_info
  rescue StandardError => e
    Rails.logger.error e.message
    flash['error'] = e.message
    redirect_to signin_path
  end

  private

  def try_to_login(email, info)
    params['back_url'] = session[:back_url]
    session.delete :back_url
    user = User.joins(:email_addresses).where(email_addresses: { address: email }).first
    if user # Existing user
      if user.registered? # Registered
        account_pending user
      elsif user.active? # Active
        handle_active_user user
        user.update_last_login_on!
      else # Locked
        handle_inactive_user user
      end
    elsif Setting.self_registration? # Create on the fly
      user = User.new
      user.mail = email
      firstname, lastname = info['name'].split if info['name'].present?
      firstname ||= info['given_name']
      lastname ||= info['family_name']
      user.firstname = firstname
      user.lastname = lastname
      user.mail = email
      login = info['login']
      login ||= info['unique_name']
      user.login = login
      user.random_password
      user.register
      case Setting.self_registration
      when '1'
        register_by_email_activation(user) do
          onthefly_creation_failed user
        end
      when '3'
        register_automatically(user) do
          onthefly_creation_failed user
        end
      else
        register_manually_by_administrator(user) do
          onthefly_creation_failed user
        end
      end
    else  # Invalid credentials
      params[:username] = email
      invalid_credentials
      raise StandardError, l(:notice_account_invalid_credentials)
    end
  end

  def oauth_client
    return @client if @client

    site = Setting.plugin_redmine_oauth[:site]&.chomp('/')
    raise StandardError, l(:oauth_invalid_provider) unless site

    @client =
      case Setting.plugin_redmine_oauth[:oauth_name]
      when 'Azure AD'
        OAuth2::Client.new(
          Setting.plugin_redmine_oauth[:client_id],
          Setting.plugin_redmine_oauth[:client_secret],
          site: site,
          authorize_url: "/#{Setting.plugin_redmine_oauth[:tenant_id]}/oauth2/authorize",
          token_url: "/#{Setting.plugin_redmine_oauth[:tenant_id]}/oauth2/token"
        )
      when 'GitLab'
        OAuth2::Client.new(
          Setting.plugin_redmine_oauth[:client_id],
          Setting.plugin_redmine_oauth[:client_secret],
          site: site,
          authorize_url: '/oauth/authorize',
          token_url: '/oauth/token'
        )
      when 'Okta'
        OAuth2::Client.new(
          Setting.plugin_redmine_oauth[:client_id],
          Setting.plugin_redmine_oauth[:client_secret],
          site: site,
          authorize_url: "/oauth2/#{Setting.plugin_redmine_oauth[:tenant_id]}/v1/authorize",
          token_url: "/oauth2/#{Setting.plugin_redmine_oauth[:tenant_id]}/v1/token"
        )
      when 'Keycloak'
        OAuth2::Client.new(
          Setting.plugin_redmine_oauth[:client_id],
          Setting.plugin_redmine_oauth[:client_secret],
          site: site,
          authorize_url: "/realms/#{Setting.plugin_redmine_oauth[:tenant_id]}/protocol/openid-connect/auth",
          token_url: "/realms/#{Setting.plugin_redmine_oauth[:tenant_id]}/protocol/openid-connect/token"
        )
      else
        raise StandardError, l(:oauth_invalid_provider)
      end
  end

  def verify_csrf_token
    if params[:state].blank? || (params[:state] != session[:oauth_csrf_token])
      render_error status: 422, message: l(:error_invalid_authenticity_token)
    end
    session.delete :oauth_csrf_token
  end
end
